File: citations/core/src/test/scala/pl/edu/icm/coansys/citations/jobs/ReferenceExtractorTest.scala
package pl.edu.icm.coansys.citations.jobs

import org.testng.annotations.Test
import com.nicta.scoobi.application.InMemoryHadoop
import pl.edu.icm.coansys.importers.models.DocumentProtos.{ReferenceMetadata, BasicMetadata, DocumentMetadata, DocumentWrapper}
import com.nicta.scoobi.Scoobi._
import org.testng.Assert._
import pl.edu.icm.coansys.citations.data.WireFormats._

/**
 * @author Mateusz Fedoryszak (m.fedoryszak@icm.edu.pl)
 */
class ReferenceExtractorTest extends InMemoryHadoop {
  def scoobiArgs = Seq()
  implicit val conf = new ScoobiConfiguration

  @Test(groups = Array("slow"))
  def basicTest() {
    val wrapperBuilder = DocumentWrapper.newBuilder()
    wrapperBuilder.setRowId("20")
    val docBuilder = DocumentMetadata.newBuilder()
    docBuilder.setKey("20")
    val metaBuilder = BasicMetadata.newBuilder()
    docBuilder.setBasicMetadata(metaBuilder)
    docBuilder.addReference(
      ReferenceMetadata.newBuilder()
        .setSourceDocKey("20")
        .setPosition(1)
        .setBasicMetadata(BasicMetadata.newBuilder()
          .setJournal("Another journal")))


    wrapperBuilder.setDocumentMetadata(docBuilder)
    inMemory {
      val in = DList(wrapperBuilder.build())
      val extracted = persist(ReferenceExtractor.extractReferences(in).materialise).asInstanceOf[Seq[(String, BasicMetadata)]]
      val elem = extracted(0)
      assertEquals(elem._1, "cit_20_1")
      assertEquals(elem._2.getJournal, "Another journal")
    }
  }
}

File: citation-matching/citation-matching-coansys-code/src/test/java/pl/edu/icm/coansys/citations/coansys/input/ReferenceExtractorTest.java
package pl.edu.icm.coansys.citations.coansys.input;

import static org.testng.Assert.assertEquals;
import static pl.edu.icm.coansys.citations.coansys.input.TestReferenceFactory.createReference;

import java.util.List;

import org.testng.annotations.Test;

import pl.edu.icm.coansys.models.DocumentProtos.BasicMetadata;
import pl.edu.icm.coansys.models.DocumentProtos.DocumentMetadata;
import pl.edu.icm.coansys.models.DocumentProtos.DocumentWrapper;
import pl.edu.icm.coansys.models.DocumentProtos.ReferenceMetadata;
import scala.Tuple2;

/**
* @author Łukasz Dumiszewski
*/

public class ReferenceExtractorTest {

    
    private ReferenceExtractor referenceExtractor = new ReferenceExtractor();
    
    
    
    //------------------------ TESTS --------------------------
    
    @Test
    public void extractReferences() {
        
        // given
        
        BasicMetadata basicMetadata = BasicMetadata.newBuilder().build();
        
        ReferenceMetadata ref1 = createReference("Raw citation");
        ReferenceMetadata ref2 = createReference("");
        ReferenceMetadata ref3 = createReference("Raw citation number 3");
        ReferenceMetadata ref4 = createReference("Raw citation 4");
        
        
        DocumentMetadata docMetadata = DocumentMetadata.newBuilder()
                                                .setKey("KEY")
                                                .setBasicMetadata(basicMetadata)
                                                .addReference(ref1).addReference(ref2).addReference(ref3).addReference(ref4)
                                                .build();
        
        DocumentWrapper docWrapper = DocumentWrapper.newBuilder().setRowId("ROW_ID").setDocumentMetadata(docMetadata).build();
        
            
        // execute
        
        List<Tuple2<String, ReferenceMetadata>> docIdReferences = referenceExtractor.extractReferences(docWrapper, 20);
        

        // assert
        
        assertEquals(2,  docIdReferences.size());
        
        assertEquals(docMetadata.getKey(), docIdReferences.get(0)._1());
        assertEquals(docMetadata.getKey(), docIdReferences.get(1)._1());
        
        assertEquals(ref1, docIdReferences.get(0)._2());
        assertEquals(ref4, docIdReferences.get(1)._2());
        
    }


        
}
